Fetch geojson using ems-client

nickpeihl opened the pull request:
## Summary

Use the `#FileLayer.getGeoJson` method from the `@elastic/ems-client` library to fetch and cache GeoJSON from Elastic Maps Service (EMS). This moves much of the topojson to geojson conversion logic to the ems-client library and reduces network calls for EMS resources from Elastic Maps.

elasticmachine commented:
Pinging @elastic/kibana-gis (Team:Geo)

nickpeihl commented on package.json:
```diff
@@ -382,7 +382,7 @@
     "tinycolor2": "1.4.1",
     "tinygradient": "0.4.3",
-    "topojson-client": "3.0.0",
+    "topojson-client": "3.1.0",
```
Match the current topojson-client version used by ems-client. Should reduce the bundle size just a little bit.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/121546/)
* Commit: 5f3ec3a999595db8ba7b269a80e82a09846e996a
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-97908/5f3ec3a999595db8ba7b269a80e82a09846e996a)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/4a8067bd58fe520a3ede059ca1f426f61cbc2215) | [after](https://github.com/elastic/kibana/commit/5f3ec3a999595db8ba7b269a80e82a09846e996a) | diff |
| --- | --- | --- | --- |
| `maps` | 772 | 771 | -1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/4a8067bd58fe520a3ede059ca1f426f61cbc2215) | [after](https://github.com/elastic/kibana/commit/5f3ec3a999595db8ba7b269a80e82a09846e996a) | diff |
| --- | --- | --- | --- |
| `maps` | 2.7MB | 2.7MB | -9.9KB |
| `regionMap` | 273.8KB | 273.9KB | +61.0B |
| total |  |  | -9.8KB |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/121546/","number":"121546","commit":"5f3ec3a999595db8ba7b269a80e82a09846e996a"}]}
PIPELINE-->

thomasneirynck commented on x-pack/plugins/maps/public/classes/sources/ems_file_source/ems_file_source.tsx:
```diff
@@ -123,11 +123,7 @@ export class EMSFileSource extends AbstractVectorSource implements IEmsFileSourc
-      emsFileLayer.getDefaultFormatType() as FORMAT_TYPE,
-      'data'
-    );
+    const featureCollection = (await emsFileLayer.getGeoJson()) ?? EMPTY_FEATURE_COLLECTION;
```
This changes the behavior slightly.

I'd throw when `getGeoJson` returns nulll, instead of returning the `EMPTY_FEATURE_COLLECTION`, to preserve the original flow. Otherwise, errors don't propagate to the layer-UX (the little warning triangle in the TOC) as they would earlier (e.g. because network connection issue, faulty map-config, ...)

thomasneirynck reviewed (commented):
thx 💯 , small comment.

thomasneirynck reviewed (approved):
code review, lgtm

![image](https://user-images.githubusercontent.com/1833023/116290443-75813380-a761-11eb-90d9-44fc91d1724a.png) always great to see

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122778/)
* Commit: a2a8d00cbd697750ffb6f836b3e73cd403897956
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-97908/a2a8d00cbd697750ffb6f836b3e73cd403897956)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/aa281ffad7c0b1808154e00b937251f36e2ff76a) | [after](https://github.com/elastic/kibana/commit/a2a8d00cbd697750ffb6f836b3e73cd403897956) | diff |
| --- | --- | --- | --- |
| `maps` | 773 | 772 | -1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/aa281ffad7c0b1808154e00b937251f36e2ff76a) | [after](https://github.com/elastic/kibana/commit/a2a8d00cbd697750ffb6f836b3e73cd403897956) | diff |
| --- | --- | --- | --- |
| `maps` | 2.7MB | 2.7MB | -9.9KB |
| `regionMap` | 273.8KB | 273.9KB | +61.0B |
| total |  |  | -9.8KB |

### History
* :green_heart: [Build #121546](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/121546/) succeeded 5f3ec3a999595db8ba7b269a80e82a09846e996a

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/121546/","number":"121546","commit":"5f3ec3a999595db8ba7b269a80e82a09846e996a"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/122778/","number":"122778","commit":"a2a8d00cbd697750ffb6f836b3e73cd403897956"}]}
PIPELINE-->

bhavyarm commented:
@nickpeihl how do I test this please? Thank you!

nickpeihl commented:
@bhavyarm Kibana stores up to ten EMS Boundaries layers in the browser's cache. We can test this by adding several of the same EMS Boundaries layer to a map in a single session. For example,

1) Open the Network tab of the browser's dev tools 
2) Add the administrative regions layer under EMS Boundaries. Note the Network tab downloads the topojson file.
3) Now add a world countries layer from EMS Boundaries. Again, the Network tab downloads a file.
4) Now add another administrative regions layer. This time, the topojson file should be loaded from the browser cache and you should not see any Network requests.

To test the cache limit we would add eleven different EMS Boundaries layers to a single map. When the eleventh layer is added, the least-recently used layer is dropped from the cache. For example,

1) In a new session, open the Network tab of the browser's dev tools 
2) Add the World countries layer from EMS Boundaries, this will be downloaded and stored in the cache.
3) Add the administrative regions layer from EMS Boundaries, this will be downloaded and stored in the cache.
4) Now add eight more different EMS Boundaries layers to the map, each of these will be downloaded and stored in the cache. The cache is now full.
5) Add another world countries layer from EMS Boundaries. There should not be additional Network request for this layer as it was cached.
6) Now add a completely different EMS Boundaries layer. This will be downloaded and stored in the cache. The tenth least-recently-used layer will be dropped from the cache. In this case, it is the administrative regions layer we added in step 3. Even though we added the world countries layer to the cache first, we used it again in step 5, so the administrative regions layer (added in step 3) is dropped.
7) Add another administrative regions layer and we should see a network request for the file which means it was not available in the cache.

Let me know if any of these doesn't make sense.

bhavyarm commented:
You rock @nickpeihl thank you!
